File: src/plugins/dash-exo-player.ts
/**
 * DashExoPlayer — explicit native DASH/HLS bridge, outside normal OTT playback.
 *
 * Android: Media3/ExoPlayer with PlayerView overlay.
 * iOS: honest {ok:false, unsupported:true} (WKWebView has no MSE; no AVPlayer DASH).
 * WebPlugin fallback for non-Cap builds only.
 * The main player never selects this plugin automatically: the native overlay
 * does not implement the complete TS player state, controls and layout contract.
 */

import { resolveNativePlugin } from "./native-bridge";
import { nativeWebFallback } from "./web-fallback";

export interface DashPlaybackState {
    duration: number;
    ended: boolean;
    error?: string;
    ok: boolean;
    playing: boolean;
    position: number;
    unsupported?: boolean;
}

export interface DashExoPlayerPlugin {
    getPlaybackState(): Promise<DashPlaybackState>;
    isDashSupported(): Promise<{ ok: boolean; unsupported?: boolean }>;
    pauseDash(): Promise<{
        ok: boolean;
        unsupported?: boolean;
        error?: string;
    }>;
    playDash(opts: {
        url: string;
        position?: number;
    }): Promise<{ ok: boolean; unsupported?: boolean; error?: string }>;
    resumeDash(): Promise<{
        ok: boolean;
        unsupported?: boolean;
        error?: string;
    }>;
    seekDash(opts: {
        position: number;
    }): Promise<{ ok: boolean; unsupported?: boolean; error?: string }>;
    stopDash(): Promise<{ ok: boolean; unsupported?: boolean; error?: string }>;
}

class DashExoPlayerWeb implements DashExoPlayerPlugin {
    getPlaybackState(): Promise<DashPlaybackState> {
        return nativeWebFallback(function () {
            return {
                duration: 0,
                ended: false,
                ok: false,
                playing: false,
                position: 0,
                unsupported: true,
            };
        });
    }

    seekDash(_opts: {
        position: number;
    }): Promise<{ ok: boolean; unsupported?: boolean }> {
        return nativeWebFallback(function () {
            return { ok: false, unsupported: true };
        });
    }

    isDashSupported(): Promise<{ ok: boolean; unsupported?: boolean }> {
        return nativeWebFallback(function () {
            console.warn(
                "[DashExoPlayer] web fallback: isDashSupported unsupported"
            );
            return { ok: false, unsupported: true };
        });
    }

    playDash(_opts: {
        url: string;
        position?: number;
    }): Promise<{ ok: boolean; unsupported?: boolean; error?: string }> {
        return nativeWebFallback(function () {
            console.warn("[DashExoPlayer] web fallback: playDash unsupported");
            return { ok: false, unsupported: true };
        });
    }

    pauseDash(): Promise<{
        ok: boolean;
        unsupported?: boolean;
        error?: string;
    }> {
        return nativeWebFallback(function () {
            console.warn("[DashExoPlayer] web fallback: pauseDash unsupported");
            return { ok: false, unsupported: true };
        });
    }

    resumeDash(): Promise<{
        ok: boolean;
        unsupported?: boolean;
        error?: string;
    }> {
        return nativeWebFallback(function () {
            console.warn(
                "[DashExoPlayer] web fallback: resumeDash unsupported"
            );
            return { ok: false, unsupported: true };
        });
    }

    stopDash(): Promise<{
        ok: boolean;
        unsupported?: boolean;
        error?: string;
    }> {
        return nativeWebFallback(function () {
            console.warn("[DashExoPlayer] web fallback: stopDash unsupported");
            return { ok: false, unsupported: true };
        });
    }
}

export const DashExoPlayer = resolveNativePlugin<DashExoPlayerPlugin>(
    "DashExoPlayer",
    () => new DashExoPlayerWeb()
);

File: ios/App/App/Plugins/DashExoPlayer.swift
import Capacitor
import Foundation

/// Honest iOS stub: WKWebView has no MSE, and this project does not ship a
/// working AVPlayer/AVFoundation DASH path. Prefer clear unsupported over
/// claiming ExoPlayer (Android-only) or faking success.
@objc(DashExoPlayer)
public class DashExoPlayer: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DashExoPlayerPlugin"
    public let jsName = "DashExoPlayer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getPlaybackState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "seekDash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isDashSupported", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pauseDash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "playDash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeDash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopDash", returnType: CAPPluginReturnPromise),
    ]

    @objc func getPlaybackState(_ call: CAPPluginCall) {
        call.resolve([
            "ok": false, "unsupported": true,
            "position": 0, "duration": 0, "playing": false, "ended": false,
        ])
    }

    @objc func seekDash(_ call: CAPPluginCall) {
        call.resolve(["ok": false, "unsupported": true])
    }

    @objc func isDashSupported(_ call: CAPPluginCall) {
        call.resolve([
            "ok": false,
            "unsupported": true,
        ])
    }

    @objc func playDash(_ call: CAPPluginCall) {
        call.resolve([
            "ok": false,
            "unsupported": true,
            "error": "DASH not supported on iOS Capacitor (no MSE; no AVPlayer DASH path)",
        ])
    }

    @objc func pauseDash(_ call: CAPPluginCall) {
        call.resolve([
            "ok": false,
            "unsupported": true,
        ])
    }

    @objc func resumeDash(_ call: CAPPluginCall) {
        call.resolve([
            "ok": false,
            "unsupported": true,
        ])
    }

    @objc func stopDash(_ call: CAPPluginCall) {
        call.resolve([
            "ok": false,
            "unsupported": true,
        ])
    }
}
